Compiler panic at tip Go due to assumption about nil basic types

Gopherjs breaks at ~[CL 284052](https://golang.org/cl/284052)~ [CL 289715](https://golang.org/cl/289715) in the Go repository, due to an assumption about nil basic types here:
https://github.com/gopherjs/gopherjs/blob/bed99a852dfeb7c8aad4ac454357a198750247e1/compiler/expressions.go#L738
(`t.Kind()` is now `types.UntypedNil`)

This is reproducible as follows:
```
> GO111MODULE=off go test -v -run=TestGoRepositoryCompilerTests ./tests                                                                                         
=== RUN   TestGoRepositoryCompilerTests                                                                             
goos: "linux", goarch: "js"                                                                                         
ok      fixedbugs/bug000.go     0.254s                                                                              
ok      fixedbugs/bug002.go     0.190s                                                                                                                                                                                                  
ok      fixedbugs/bug003.go     0.186s                                                                              
ok      fixedbugs/bug004.go     0.176s                                                                                                                                                                                                  
ok      fixedbugs/bug005.go     0.178s                                                                              
# go run run.go -- fixedbugs/bug006.go                                                                                                                                                                                                  
exit status 2                                                                                                       
panic: unexpected basic type [recovered]                                                                            
        panic: unexpected basic type                                                                                
                                                                                                                    
goroutine 1 [running]:                                                                                              
go/types.(*Checker).handleBailout(0xc00024a000, 0xc0000eed00)                                                       
        /usr/local/google/home/rfindley/src/go2/src/go/types/check.go:245 +0x9f                                   
panic(0x918020, 0xa87510)                                                                                       
        /usr/local/google/home/rfindley/src/go2/src/runtime/panic.go:965 +0x1c7
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateExpr(0xc00052c2c0, 0xa9a310, 0xc00071c1c0, 0xa9a310)
        /usr/local/google/home/rfindley/dev/go/src/github.com/gopherjs/gopherjs/compiler/expressions.go:739 +0xf0d3
```

This appears to be a case of [Hyrum's law](https://www.hyrumslaw.com/) -- I'm not aware of documentation stating that a basic nil must be `types.UnsafePointer`, though please let me know if I've missed something.

Unless there is evidence that this implicit assumption about nil basic types is widespread, we'd like to land this change in `go/types` for go1.17.  It seems like a trivial fix, but I'm not familiar with the gopherjs compiler. Could someone take a look? Thanks.

CC @dmitshur

Sorry, I linked the wrong CL above (will amend): https://golang.org/cl/284052 is the compiler change, https://golang.org/cl/289715 is the go/types port.

Yes, the example in  https://github.com/gopherjs/gopherjs/issues/1011#issuecomment-810533541 is one example of what is changing. Note that 'nil' in the following was already reported as untyped nil:

```
var _ *int = nil
```

See more at https://golang.org/issue/13061.  Here are playground examples illustrating the change (I'm using go2goplay only because it has the new logic; none of this pertains directly to generics):
https://play.golang.org/p/LroExeoXjg-
https://go2goplay.golang.org/p/LroExeoXjg-

The actual source that causes the panic is here:

https://github.com/golang/go/blob/46cb016190389b7e37b21f04e5343a628ca1f662/src/io/multi.go#L72

Here's an error message with more information:
```
# go run run.go -- fixedbugs/bug006.go                                                                              
exit status 2                                                                                                       
panic: unexpected basic type 25@/usr/local/google/home/rfindley/sdk/go1.12.17/src/io/multi.go:72:37 [recovered]
```

You're right, it's not completely trivial. I assumed it was straightforward because there must have been pre-existing handling for assignments, for example:
https://github.com/gopherjs/gopherjs/blob/bed99a852dfeb7c8aad4ac454357a198750247e1/compiler/expressions.go#L1120

Here is more of the panicking stack:
```
# go run run.go -- fixedbugs/bug006.go                                                                              
exit status 2                                                                                                       
panic: unexpected basic type 25@/usr/local/google/home/rfindley/sdk/go1.12.17/src/io/multi.go:72:37 [recovered]
        panic: unexpected basic type 25@/usr/local/google/home/rfindley/sdk/go1.12.17/src/io/multi.go:72:37         
                                                                                                                    
goroutine 1 [running]:                                                                                              
go/types.(*Checker).handleBailout(0xc0002ca000, 0xc0008e4d00)                      
        /usr/local/google/home/rfindley/src/go2/src/go/types/check.go:245 +0x9f                                     
panic(0x918020, 0xc000487950)                                                                                       
        /usr/local/google/home/rfindley/src/go2/src/runtime/panic.go:965 +0x1c7                                   
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateExpr(0xc0004322c0, 0xa9a4f0, 0xc000280700, 0xa9a4f0) 
        /usr/local/google/home/rfindley/dev/go/src/github.com/gopherjs/gopherjs/compiler/expressions.go:739 +0xf289 
github.com/gopherjs/gopherjs/compiler.(*funcContext).formatExprInternal.func3.1(0x0, 0x0)                         
        /usr/local/google/home/rfindley/dev/go/src/github.com/gopherjs/gopherjs/compiler/expressions.go:1293 +0xce  
github.com/gopherjs/gopherjs/compiler.(*funcContext).formatExprInternal.func3(0x6500, 0x0)                         
        /usr/local/google/home/rfindley/dev/go/src/github.com/gopherjs/gopherjs/compiler/expressions.go:1314 +0x28c
github.com/gopherjs/gopherjs/compiler.(*funcContext).formatExprInternal.func1(0xc0008e00c0)                       
        /usr/local/google/home/rfindley/dev/go/src/github.com/gopherjs/gopherjs/compiler/expressions.go:1248 +0xc2  
github.com/gopherjs/gopherjs/compiler.(*funcContext).formatExprInternal(0xc0004322c0, 0x9d2c59, 0x22, 0xc0008e0358, 0x2, 0x2, 0x0, 0x438800)
        /usr/local/google/home/rfindley/dev/go/src/github.com/gopherjs/gopherjs/compiler/expressions.go:1287 +0x2ad
github.com/gopherjs/gopherjs/compiler.(*funcContext).formatExpr(...)                                        
        /usr/local/google/home/rfindley/dev/go/src/github.com/gopherjs/gopherjs/compiler/expressions.go:1228        
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateConversion(0xc0004322c0, 0xa9a4f0, 0xc000280700, 0xa95cf0, 0xc0003761c0, 0x0)
        /usr/local/google/home/rfindley/dev/go/src/github.com/gopherjs/gopherjs/compiler/expressions.go:1075 +0x13b4
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateExpr(0xc0004322c0, 0xa9a1f0, 0xc000332440, 0xc0000c2910)
        /usr/local/google/home/rfindley/dev/go/src/github.com/gopherjs/gopherjs/compiler/expressions.go:542 +0x32b
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateStmt(0xc0004322c0, 0xa9a070, 0xc000352d00, 0x0)    
```

--- a/README.md
+++ b/README.md
@@ -68,10 +68,11 @@ If you include an argument, it will be the root from which everything is served.
 
 There is one GopherJS-specific environment variable:
 
-```
-GOPHERJS_GOROOT - if set, GopherJS uses this value as the default GOROOT value,
-                  instead of using the system GOROOT as the default GOROOT value
-```
+ - `GOPHERJS_GOROOT` - if set, GopherJS uses this value as the default GOROOT
+   value, instead of using the system GOROOT as the default GOROOT value
+ - `GOPHERJS_SKIP_VERSION_CHECK` - if set to true, GopherJS will not check 
+   Go version in the GOROOT for compatibility with the GopherJS release. This
+	 is primarily useful for testing GopherJS against unreleased versions of Go.
 
 ### Performance Tips
 

--- a/compiler/expressions.go
+++ b/compiler/expressions.go
@@ -1069,25 +1069,48 @@ func (fc *funcContext) translateConversion(expr ast.Expr, desiredType types.Type
 		}
 
 	case *types.Pointer:
+		if types.Identical(exprType, types.Typ[types.UntypedNil]) {
+			// Fall through to the fc.translateImplicitConversionWithCloning(), which
+			// handles conversion from untyped nil to a pointer type.
+			break
+		}
+
 		switch u := t.Elem().Underlying().(type) {
-		case *types.Array:
+		case *types.Array: // (*[N]T)(expr) — converting expr to a pointer to an array.
 			return fc.translateExpr(expr)
-		case *types.Struct:
+		case *types.Struct: // (*StructT)(expr) — converting expr to a pointer to a struct.
 			if fc.pkgCtx.Pkg.Path() == "syscall" && types.Identical(exprType, types.Typ[types.UnsafePointer]) {
+				// Special case: converting an unsafe pointer to a byte array into a
+				// struct pointer when handling syscalls.
+				// TODO(nevkontakte): Add a runtime assertion that the unsafe.Pointer is
+				// indeed pointing at a byte array.
 				array := fc.newVariable("_array")
 				target := fc.newVariable("_struct")
 				return fc.formatExpr("(%s = %e, %s = %e, %s, %s)", array, expr, target, fc.zeroValue(t.Elem()), fc.loadStruct(array, target, u), target)
 			}
+			// Convert between structs of different types but identical layouts,
+			// for example:
+			// type A struct { foo int }; type B A; var a *A = &A{42}; var b *B = (*B)(a)
+			//
+			// TODO(nevkontakte): Should this only apply when exprType is a pointer to a
+			// struct as well?
 			return fc.formatExpr("$pointerOfStructConversion(%e, %s)", expr, fc.typeName(t))
 		}
 
-		if !types.Identical(exprType, types.Typ[types.UnsafePointer]) {
-			exprTypeElem := exprType.Underlying().(*types.Pointer).Elem()
-			ptrVar := fc.newVariable("_ptr")
-			getterConv := fc.translateConversion(fc.setType(&ast.StarExpr{X: fc.newIdent(ptrVar, exprType)}, exprTypeElem), t.Elem())
-			setterConv := fc.translateConversion(fc.newIdent("$v", t.Elem()), exprTypeElem)
-			return fc.formatExpr("(%1s = %2e, new %3s(function() { return %4s; }, function($v) { %1s.$set(%5s); }, %1s.$target))", ptrVar, expr, fc.typeName(desiredType), getterConv, setterConv)
+		if types.Identical(exprType, types.Typ[types.UnsafePointer]) {
+			// TODO(nevkontakte): Why do we fall through to the implicit conversion here?
+			// Conversion from unsafe.Pointer() requires explicit type conversion: https://play.golang.org/p/IQxtmpn1wgc.
+			// Possibly related to https://github.com/gopherjs/gopherjs/issues/1001.
+			break // Fall through to fc.translateImplicitConversionWithCloning() below.
 		}
+		// Handle remaining cases, for example:
+		// type iPtr *int; var c int = 42; println((iPtr)(&c));
+		// TODO(nevkontakte): Are there any other cases that fall into this case?
+		exprTypeElem := exprType.Underlying().(*types.Pointer).Elem()
+		ptrVar := fc.newVariable("_ptr")
+		getterConv := fc.translateConversion(fc.setType(&ast.StarExpr{X: fc.newIdent(ptrVar, exprType)}, exprTypeElem), t.Elem())
+		setterConv := fc.translateConversion(fc.newIdent("$v", t.Elem()), exprTypeElem)
+		return fc.formatExpr("(%1s = %2e, new %3s(function() { return %4s; }, function($v) { %1s.$set(%5s); }, %1s.$target))", ptrVar, expr, fc.typeName(desiredType), getterConv, setterConv)
 
 	case *types.Interface:
 		if types.Identical(exprType, types.Typ[types.UnsafePointer]) {
@@ -1185,6 +1208,7 @@ func (fc *funcContext) loadStruct(array, target string, s *types.Struct) string 
 		case *types.Array:
 			code += fmt.Sprintf(`, %s = new ($nativeArray(%s))(%s.buffer, $min(%s.byteOffset + %d, %s.buffer.byteLength))`, field.Name(), typeKind(t.Elem()), array, array, offsets[i], array)
 		}
+		// TODO(nevkontakte): Explicitly panic if unsupported field type is encountered?
 	}
 	return code
 }

--- a/compiler/package.go
+++ b/compiler/package.go
@@ -120,7 +120,20 @@ func (pi packageImporter) Import(path string) (*types.Package, error) {
 	return pi.importContext.Packages[a.ImportPath], nil
 }
 
-func Compile(importPath string, files []*ast.File, fileSet *token.FileSet, importContext *ImportContext, minify bool) (*Archive, error) {
+func Compile(importPath string, files []*ast.File, fileSet *token.FileSet, importContext *ImportContext, minify bool) (_ *Archive, err error) {
+	defer func() {
+		e := recover()
+		if e == nil {
+			return
+		}
+		if fe, ok := bailingOut(e); ok {
+			// Orderly bailout, return whatever clues we already have.
+			err = fe
+			return
+		}
+		// Some other unexpected panic, catch the stack trace and return as an error.
+		err = bailout(e)
+	}()
 	typesInfo := &types.Info{
 		Types:      make(map[ast.Expr]types.TypeAndValue),
 		Defs:       make(map[*ast.Ident]types.Object),

--- a/compiler/statements.go
+++ b/compiler/statements.go
@@ -22,6 +22,27 @@ func (fc *funcContext) translateStmtList(stmts []ast.Stmt) {
 }
 
 func (fc *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
+	defer func() {
+		err := recover()
+		if err == nil {
+			return
+		}
+		if _, yes := bailingOut(err); yes {
+			panic(err) // Continue orderly bailout.
+		}
+
+		// Oh noes, we've tried to compile something so bad that compiler paniced
+		// and ran away. Let's gather some debugging clues.
+		bail := bailout(err)
+		pos := stmt.Pos()
+		if fc.posAvailable && fc.pos.IsValid() {
+			pos = fc.pos
+		}
+		fmt.Fprintf(bail, "Occurred while compiling statement at %s:\n", fc.pkgCtx.fileSet.Position(pos))
+		ast.Fprint(bail, fc.pkgCtx.fileSet, stmt, ast.NotNilFilter)
+		panic(bail) // Initiate orderly bailout.
+	}()
+
 	fc.SetPos(stmt.Pos())
 
 	stmt = filter.IncDecStmt(stmt, fc.pkgCtx.Info.Info)

--- a/compiler/statements.go
+++ b/compiler/statements.go
@@ -386,7 +386,7 @@ func (fc *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 		case len(s.Lhs) == 1 && len(s.Rhs) == 1:
 			lhs := astutil.RemoveParens(s.Lhs[0])
 			if isBlank(lhs) {
-				fc.Printf("$unused(%s);", fc.translateExpr(s.Rhs[0]))
+				fc.Printf("$unused(%s);", fc.translateImplicitConversion(s.Rhs[0], fc.pkgCtx.TypeOf(s.Lhs[0])))
 				return
 			}
 			fc.Printf("%s", fc.translateAssign(lhs, s.Rhs[0], s.Tok == token.DEFINE))

--- a/compiler/utils.go
+++ b/compiler/utils.go
@@ -3,12 +3,14 @@ package compiler
 import (
 	"bytes"
 	"encoding/binary"
+	"errors"
 	"fmt"
 	"go/ast"
 	"go/constant"
 	"go/token"
 	"go/types"
 	"net/url"
+	"runtime/debug"
 	"sort"
 	"strconv"
 	"strings"
@@ -737,4 +739,56 @@ func (st signatureTypes) Param(i int, ellipsis bool) types.Type {
 // ErrorAt annotates an error with a position in the source code.
 func ErrorAt(err error, fset *token.FileSet, pos token.Pos) error {
 	return fmt.Errorf("%s: %w", fset.Position(pos), err)
+}
+
+// FatalError is an error compiler panics with when it encountered a fatal error.
+//
+// FatalError implements io.Writer, which can be used to record any free-form
+// debugging details for human consumption. This information will be included
+// into String() result along with the rest.
+type FatalError struct {
+	cause interface{}
+	stack []byte
+	clues strings.Builder
+}
+
+func (b FatalError) Unwrap() error {
+	if b.cause == nil {
+		return nil
+	}
+	if err, ok := b.cause.(error); ok {
+		return err
+	}
+	if s, ok := b.cause.(string); ok {
+		return errors.New(s)
+	}
+	return fmt.Errorf("[%T]: %v", b.cause, b.cause)
+}
+
+// Write implements io.Writer and can be used to store free-form debugging clues.
+func (b *FatalError) Write(p []byte) (n int, err error) { return b.clues.Write(p) }
+
+func (b FatalError) Error() string {
+	buf := &strings.Builder{}
+	fmt.Fprintln(buf, "[compiler panic] ", b.Unwrap())
+	if b.clues.Len() > 0 {
+		fmt.Fprintln(buf, "\n", b.clues.String())
+	}
+	if len(b.stack) > 0 {
+		fmt.Fprintln(buf, "\n", string(b.stack))
+	}
+	return buf.String()
+}
+
+func bailout(cause interface{}) *FatalError {
+	b := &FatalError{
+		cause: cause,
+		stack: debug.Stack(),
+	}
+	return b
+}
+
+func bailingOut(err interface{}) (*FatalError, bool) {
+	fe, ok := err.(*FatalError)
+	return fe, ok
 }

--- a/compiler/version_check.go
+++ b/compiler/version_check.go
@@ -6,7 +6,9 @@ import (
 	"bytes"
 	"fmt"
 	"io/ioutil"
+	"os"
 	"path/filepath"
+	"strconv"
 )
 
 // Version is the GopherJS compiler version string.
@@ -19,6 +21,9 @@ const GoVersion = 16
 // at goroot, and reports an error if it's not compatible
 // with this version of the GopherJS compiler.
 func CheckGoVersion(goroot string) error {
+	if nvc, err := strconv.ParseBool(os.Getenv("GOPHERJS_SKIP_VERSION_CHECK")); err == nil && nvc {
+		return nil
+	}
 	v, err := ioutil.ReadFile(filepath.Join(goroot, "VERSION"))
 	if err != nil {
 		return fmt.Errorf("GopherJS %s requires a Go 1.16.x distribution, but failed to read its VERSION file: %v", Version, err)

--- a/tests/misc_test.go
+++ b/tests/misc_test.go
@@ -712,7 +712,8 @@ func TestReflectMapIterationAndDelete(t *testing.T) {
 }
 
 func TestUntypedNil(t *testing.T) {
-	// This test makes sure GopherJS compiler correctly handles untyped nil.
+	// This test makes sure GopherJS compiler is able to correctly infer the
+	// desired type of an untyped nil ident.
 	// See https://github.com/gopherjs/gopherjs/issues/1011 for details.
 
 	// Code below is based on test cases from https://golang.org/cl/284052.

--- a/tests/misc_test.go
+++ b/tests/misc_test.go
@@ -710,3 +710,136 @@ func TestReflectMapIterationAndDelete(t *testing.T) {
 		t.Fatalf("got %d, want %d", got, want)
 	}
 }
+
+func TestUntypedNil(t *testing.T) {
+	// This test makes sure GopherJS compiler correctly handles untyped nil.
+	// See https://github.com/gopherjs/gopherjs/issues/1011 for details.
+
+	// Code below is based on test cases from https://golang.org/cl/284052.
+	var _ *int = nil
+	var _ func() = nil
+	var _ []byte = nil
+	var _ map[int]int = nil
+	var _ chan int = nil
+	var _ interface{} = nil
+
+	{
+		var (
+			x *int = nil
+			_      = x
+		)
+	}
+	{
+		var (
+			x func() = nil
+			_        = x
+		)
+	}
+	{
+		var (
+			x []byte = nil
+			_        = x
+		)
+	}
+	{
+		var (
+			x map[int]int = nil
+			_             = x
+		)
+	}
+	{
+		var (
+			x chan int = nil
+			_          = x
+		)
+	}
+	{
+		var (
+			x interface{} = nil
+			_             = x
+		)
+	}
+
+	{
+		var (
+			x *int
+			_ = x == nil
+		)
+	}
+	{
+		var (
+			x func()
+			_ = x == nil
+		)
+	}
+	{
+		var (
+			x []byte
+			_ = x == nil
+		)
+	}
+	{
+		var (
+			x map[int]int
+			_ = x == nil
+		)
+	}
+	{
+		var (
+			x chan int
+			_ = x == nil
+		)
+	}
+	{
+		var (
+			x interface{}
+			_ = x == nil
+		)
+	}
+	var _ = (*int)(nil)
+	var _ = (func())(nil)
+	var _ = ([]byte)(nil)
+	var _ = (map[int]int)(nil)
+	var _ = (chan int)(nil)
+	var _ = (interface{})(nil)
+	{
+		f := func(*int) {}
+		f(nil)
+	}
+	{
+		f := func(func()) {}
+		f(nil)
+	}
+	{
+		f := func([]byte) {}
+		f(nil)
+	}
+	{
+		f := func(map[int]int) {}
+		f(nil)
+	}
+	{
+		f := func(chan int) {}
+		f(nil)
+	}
+	{
+		f := func(interface{}) {}
+		f(nil)
+	}
+	{
+		f := func(*int) {}
+		f(nil)
+	}
+	{
+		f := func(*int) {}
+		f(nil)
+	}
+	{
+		f := func(*int) {}
+		f(nil)
+	}
+	{
+		f := func(*int) {}
+		f(nil)
+	}
+}

